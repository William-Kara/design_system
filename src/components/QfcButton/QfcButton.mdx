import { ArgTypes, Meta } from '@storybook/blocks';
import { Story } from '@storybook/addon-docs';
import QfcButton from '@/components/QfcButton/QfcButton.vue';

<Meta title="Elements/Buttons/Button" />

# Definition

The QfcButton is clickable element which can be used in forms, or anywhere that needs simple validation functionality.

This button may display text with ou without icons and has different appearances. Use a "Primary" button to call attention to the main action. Don't use multiple primary buttons on one page and use a "Secondary" button for secondary actions, such as â€œCancel".

## Props

The following props are available for the QfcButton component:

<ArgTypes of={QfcButton} />

# Usage

To use QfcButton, style needs to be imported separately.

```sass
@import '@website/design-system-vuejs/dist/styles/bulma/elements/button.scss';
@import '@website/design-system-vuejs/dist/styles/components/QfcButton.scss';
```

```html
<template>
  <QfcButton label="Button1" isDisabled="true" secondary="true" icon="star" />
</template>
```

# Declinaison - Primary

## Button

<Story name="Primary" id="elements-buttons-button--primary" />

## Button with icon

<Story name="Primary" id="elements-buttons-button--primary-with-icon" />

## Button - Disabled

<Story name="Primary" id="elements-buttons-button--primary-disabled" />

# Declinaison - Secondary

## Button

<Story name="Primary" id="elements-buttons-button--secondary" />

## Button with icon

<Story name="Primary" id="elements-buttons-button--secondary-with-icon" />

## Button - Disabled

<Story name="Primary" id="elements-buttons-button--secondary-disabled" />
