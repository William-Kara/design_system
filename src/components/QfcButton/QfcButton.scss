.qfc-btn {
  position: relative;
  transition: $transition-time;
  border-radius: $br-xs;
  background-color: $primary;
  border: $primary;

  &.qfc-btn--lg {
    font-size: $fs-m;
    height: 3.5rem;
  }
  &.qfc-btn--md {
    font-size: $fs-m;
    height: 2.8rem;
  }
  &.qfc-btn--sm {
    font-size: $fs-s;
    height: 2.3rem;
  }

  &.qfc-btn--secondary {
    background-color: $background;
    border: 1px solid $border;
    color: $text;
    &:hover:enabled {
      color: $text;
      background-color: $hover-dark;
    }
    &:focus {
      border: 1px solid white;
      outline-offset: 0px;
      outline: 3px solid $secondary;
    }
  }
  &__icon {
    padding-right: 0.25rem;
  }
  &:after {
    font-family: 'Material Icons Outlined';
    content: 'trending_flat';
    position: absolute;
    opacity: 0;
    //right: -20px;
    left: calc(100% - 2.3rem);
    transition: $transition-time-fast;
  }
  &:hover:enabled {
    padding-left: 1.5rem;
    padding-right: 2.5rem;

    &::after {
      opacity: 1;
      right: 1rem;
    }
  }
  &:active {
    background-color: $primary-active;
  }
  &:focus {
    outline: 1px solid $primary;
    outline-offset: 2px;
  }
  &:disabled {
    color: $grey-alto;
    background-color: $disabled;
    opacity: 1;
    &:after {
      content: '';
    }
  }
}
