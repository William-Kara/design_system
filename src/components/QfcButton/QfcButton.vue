<script lang="ts" setup>
const sizeType = ['small', 'medium', 'large'] as const;
type SizeType = (typeof sizeType)[number];
interface Props {
  isDisabled?: Boolean;
  label: string;
  secondary?: Boolean;
  icon?: String;
  size: SizeType;
}
withDefaults(defineProps<Props>(), {
  isDisabled: false,
  label: 'Label',
  secondary: false,
  size: 'large',
});
</script>
<template>
  <button
    ref="btn"
    :class="{
      button: true,
      'qfc-btn': true,
      'qfc-btn--lg': size === 'large',
      'qfc-btn--md': size === 'medium',
      'qfc-btn--sm': size === 'small',
      'qfc-btn--secondary': secondary,
    }"
    :title="label"
    :disabled="isDisabled"
    :aria-disabled="isDisabled"
  >
    <span v-if="icon" class="qfc-btn__icon material-icons">{{ icon }}</span>
    {{ label }}
  </button>
</template>

<style lang="scss" scoped src="./QfcButton.scss" />
