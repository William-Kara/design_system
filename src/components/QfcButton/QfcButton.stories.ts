import QfcButton from './QfcButton.vue';
import type { Meta, StoryObj } from '@storybook/vue3';

const meta: Meta<typeof QfcButton> = {
  component: QfcButton,
  title: 'Elements/Buttons/Button',
  argTypes: {
    label: {
      type: { name: 'string', required: true },
      defaultValue: 'Label',
    },
    icon: {
      type: { name: 'string', required: false },
      defaultValue: '',
    },
    secondary: {
      type: { name: 'boolean', required: false },
    },
    isDisabled: {
      type: { name: 'boolean', required: false },
      defaultValue: false,
    },
  },
};
export default meta;

type Story = StoryObj<typeof meta>;
export const Primary: Story = {
  args: {
    label: 'Primary Button',
    isDisabled: false,
    secondary: false,
  },
};

export const PrimaryWithIcon: Story = {
  args: {
    ...Primary.args,
    icon: 'star',
  },
};

export const PrimaryDisabled: Story = {
  args: {
    ...Primary.args,
    isDisabled: true,
  },
};

export const Secondary: Story = {
  args: {
    ...Primary.args,
    label: 'Secondary Button',
    secondary: true,
    size: 'medium',
  },
};

export const SecondaryWithIcon: Story = {
  args: {
    ...Secondary.args,
    icon: 'star',
  },
};

export const SecondaryDisabled: Story = {
  args: {
    ...Secondary.args,
    isDisabled: true,
    size: 'small',
  },
};
