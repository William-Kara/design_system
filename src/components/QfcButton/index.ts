import QfcButton from './QfcButton.vue';

export { QfcButton };
export default QfcButton;
