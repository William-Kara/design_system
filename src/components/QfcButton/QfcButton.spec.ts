import { it, describe, beforeEach, expect } from 'vitest';

import { shallowMount } from '@vue/test-utils';
import QfcButton from './QfcButton.vue';

let wrapper: any = null;

const propsValues = {
  label: 'Text1',
  isDisabled: false,
  secondary: false,
  size: 'large',
};

beforeEach(() => {
  wrapper = shallowMount(QfcButton, {
    propsData: propsValues,
  });
});

describe('QfcButton Component', () => {
  it('should mount properly', () => {
    expect(wrapper).toBeTruthy();
  });

  it('should have correct props', () => {
    expect(wrapper.props()).toEqual(propsValues);
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
