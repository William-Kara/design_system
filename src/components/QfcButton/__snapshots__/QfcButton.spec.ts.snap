// Vitest Snapshot v1

exports[`QfcButton Component > should render correctly 1`] = `
"<button data-v-f954cdd1=\\"\\" class=\\"button qfc-btn qfc-btn--lg\\" title=\\"Text1\\" aria-disabled=\\"false\\">
  <!--v-if--> Text1
</button>"
`;
